pubsub/pubsub.go

```diff
@@ -113,13 +115,30 @@ func mergeSubscriberCallOptions(a *vkit.SubscriberCallOptions, b *vkit.Subscribe
 	return res
 }
 
+// DetectProjectID is a sentinel value that instructs NewClient to detect the
+// project ID. It is given in place of the projectID argument. NewClient will
+// use the project ID from the given credentials or the default credentials
+// (https://developers.google.com/accounts/docs/application-default-credentials)
+// if no credentials were provided. When providing credentials, not all
+// options will allow NewClient to extract the project ID. Specifically a JWT
+// does not have the project ID encoded.
+const DetectProjectID = "*detect-project-id*"
+
+// ErrEmptyProjectID denotes that the project string passed into NewClient was empty.
+// Please provide a valid project ID or use the DetectProjectID sentinel value to detect
+// project ID from well defined sources.
+var ErrEmptyProjectID = errors.New("pubsub: projectID string is empty")
+
 // NewClient creates a new PubSub client. It uses a default configuration.
 func NewClient(ctx context.Context, projectID string, opts ...option.ClientOption) (c *Client, err error) {
 	return NewClientWithConfig(ctx, projectID, nil, opts...)
 }
 
 // NewClientWithConfig creates a new PubSub client.
 func NewClientWithConfig(ctx context.Context, projectID string, config *ClientConfig, opts ...option.ClientOption) (c *Client, err error) {
+	if projectID == "" {
```

Comment:
@hongalex doesn't having this check early on  break the `opts.WithoutAuthentication` case since it now requires you to pass a dummy projectID to instantiate the clients.

Reply:
Isn't this a breaking change that warrants a major version bump instead of a minor bump?

Reply:
While this may be backwards incompatible, but falls under the realm of a bug fix and doesn't warrant a major version bump. Passing in an empty string to the client is unsupported behavior and should have resulted in an error, even when using unauthenticated. The project string is used to construct paths like "projects/my-project/topics/my-topic" and creating resources like "projects//topics/my-topic" is undefined behavior.